core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -49,14 +49,270 @@
 #ifdef KOKKOS_ENABLE_CUDA
 #include <cuda_fp16.h>
 #include <cstdint>
+#include <cstdio>
 
 #ifndef KOKKOS_IMPL_HALF_TYPE_DEFINED
 // Make sure no one else tries to define half_t
 #define KOKKOS_IMPL_HALF_TYPE_DEFINED
 
 namespace Kokkos {
 namespace Experimental {
-using half_t                       = __half;
+
+// Forward declarations
+class half_t;
+
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(float val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(double val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(short val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(int val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(long val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(long long val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned short val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned int val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long val);
+KOKKOS_INLINE_FUNCTION
+half_t cast_to_half(unsigned long long val);
+
+class half_t {
+  __half val;
+
+public:
+
+  // Conversion operator for __half(bar) = half_t(foo)
+  KOKKOS_FUNCTION
+  operator __half() const {
+    return val;
+  }
+
+// Changing below to 1 produces the constructor overload error
+#if 0
+  // Conversion operator for bool(bar) = half_t(foo)
+  KOKKOS_FUNCTION
+  operator bool() const {
+    return static_cast<bool>(__half2float(val) == 0.0F);
+  }
+#endif
+
+  
+  KOKKOS_FUNCTION
+  half_t(__half rhs = 0):val(rhs) {}
+
+  // Cast rhs to half for assignment to lhs of type half_t
+  template <class T>
+  KOKKOS_FUNCTION
+  half_t(T rhs):val(cast_to_half(rhs)) {}
```

Comment:
the problem with the above bool conversion put in goes away if we do constructor forwarding here:
```
half_t(T rhs):half_t(cast_to_half(rhs)) {}
```